Source/Core/VideoBackends/D3D12/Render.cpp

```diff
@@ -1244,6 +1242,8 @@ void Renderer::SetSamplerState(int stage, int tex_index, bool custom_tex)
 	if (tex_index)
 		stage += 4;
 
+	const auto original_hex = gx_state.sampler[stage].hex;
```

Comment:
> It's possible to add bool support to BitField by separating ...

While this is indeed a good suggestion, this still doesn't solve the issue that it's very easy to accomplish undefined behavior. BitField is inherently fragile like that when used with unions.

> I was just curious about what the alternative design for BPMemory would look like...

BPMemory would remain as a struct. Just it's constituent members wouldn't leverage BitField (or C bitfields).

For example:


```cpp
union IND_MTXA
{
    struct
    {
        s32 ma : 11;
        s32 mb : 11;
        u32 s0 : 2; // bits 0-1 of scale factor
        u32 rid : 8;
    };
    u32 hex;
};
```

would be

```cpp
class IND_MTXA final
{
public:
    // Just here for example. Likely not necessary.
    constexpr IND_MTXA() = default;
    constexpr explicit IND_MTXA(u32 value) : m_value(value) {}

    IND_MTXA& operator=(u32 value)
    {
        m_value = value;
        return *this;
    }

    // Bit arith can be simplified/made less 'boilerplatey' and checked at
    // compile-time by having a templated constexpr free function in
    // common for bit extraction, which can provide static asserts for other
    // desirable things (prevent shifting into sign bit for signed types, etc).

    constexpr u32 MA() const { return /* bit arith here extracting bits */ }
    // etc (we don't care about rid so it doesn't need to be included).

private:
    u32 m_value = 0;
};